waku-org/nwaku
# feat(cbindings): adding tiny Waku Relay example in Golang

# Description
Nwaku integration library (libwaku.so) in Golang.

# Changes

- [x] Adding new waku.go that shows an example on how to integrate nwaku (libwaku.so) in Golang.
- [x] Adding README.md file to explain how to use the example.

## How to test
Golang should be installed beforehand. Tested with `go version go1.20.5 linux/amd64`.

1. Open one terminal from the repo's root folder.
  - ```code
      make libwaku
      ```
  - ```code
    export LD_LIBRARY_PATH=build
    ```
  - ```code
    go run examples/golang/waku.go -key 364d111d729a6eb6d2e6113e163f017b5ef03a6f94c9b5b7bb1bb36fa5cb07a9
    ```

2. Open a second terminal from the repo's root folder.
  - ```code
    export LD_LIBRARY_PATH=build
    ```
  - ```code
    ./build/wakunode2 --config-file=cfg_node_b.txt
    ```        
[cfg_node_b.txt](https://github.com/waku-org/nwaku/files/11725956/cfg_node_b.txt)

3. Open a third terminal from the repo's root folder.
  - ```code
    while [ true ]; do curl -d "{\"jsonrpc\":\"2.0\",\"id\":"$(date +%s)",\"method\":\"post_waku_v2_relay_v1_message\", \"params\":[\"/waku/2/default-waku/proto\", {\"timestamp\":"$(date +%s)", \"payload\":\"VGhpcyBpcyBhIG1lc3NhZ2UK\"}]}" --header "Content-Type: application/json" http://localhost:8546; done
    ```
    
From the first terminal (Golang), it should appear perioric logs like:

`Event received:  {"pubsubTopic":"/waku/2/default-waku/proto","messageId":"TODO","wakuMessage":{"payload":"This is a message\n","contentTopic":"/waku/2/default-content/proto","version":0,"timestamp":1686589489},"eventType":"message"}`

## Issue
https://github.com/waku-org/nwaku/issues/1632
